### ISSUE TITLE
No effect on fans for XPS 9315 on Ubuntu 23.10

### ISSUE THREAD

**Issue by tova-cup** [Wer-Wolf/i8kutils#4]
Hi @Wer-Wolf ,

Thank you for creating this fork of i8kutils and your work on the linux community!

I've been trying to control the fans on my XPS 13 but to no avail.
I tried both the vitorafsr version of i8kutils and your fork but the fans doesn't seem to react in both cases.

In particuliar, running `i8kctl` on the original i8kutlils returns :
`1.0 1.1 CXGGY04 -1 -1 -1 -1 -1 -1 -1`

However running your fork doesn't return any output.
Of note, I have successfully loaded the i8k kernel module with `modprobe dell-smm-hwmon ignore_dmi=1` and disabled secure boot.

I have noticed however that running `dellfan` doesn't give any result either despite the fact that the fans are detected by `lm-sensors`.

I'm a bit lost to be honest, do you have any ideas ?

**Comment by Wer-Wolf** [comment]
Does the path `/sys/bus/wmi/devices/F1DDEE52-063C-4784-A11E-8A06684B9B01` exist on your machine?

If yes, then your machine seems to use a different interface for calling the SMM handler for fan control,
support for which will appear in kernel 6.8.

**Comment by tova-cup** [comment]
Thanks for your answer !

Indeed said path exists ; while waiting for the release of the 6.8 kernel for ubuntu, is there a way to retrieve a module I can load with `insmod` in the meantime ?

**Comment by Wer-Wolf** [comment]
[dell-smm-hwmon.zip](https://github.com/Wer-Wolf/i8kutils/files/14547198/dell-smm-hwmon.zip)

You can try to build this module using `make` and load it using `sudo insmod dell-smm-hwmon.ko` (unload the old driver first with `sudo modprobe -r dell-smm-hwmon`).

**Comment by tova-cup** [comment]
Thanks a lot !

I did load your module and installed your fork of `i8kutils`, and when I run `i8kctl` I get the following output :
`fan 1 state: 0
temperature 1: 37 °C
temperature 2: 30 °C
temperature 3: 30 °C
temperature 4: 30 °C
temperature 5: 29 °C
temperature 6: 28 °C
temperature 7: 27 °C
temperature 8: 26 °C
`

However, trying to control the fan has no effect (for instance after running `sudo i8kctl fan1 2`)

**Comment by Wer-Wolf** [comment]
Seems that we might need to whitelist your machine for being able to disable BIOS fan control.

Can you send me the output of `dmidecode`?

**Comment by tova-cup** [comment]
Yes of of course.

Here it is : [dmidecode.log](https://github.com/Wer-Wolf/i8kutils/files/14547271/dmidecode.log)

**Comment by Wer-Wolf** [comment]
[dell-smm-hwmon.zip](https://github.com/Wer-Wolf/i8kutils/files/14547537/dell-smm-hwmon.zip)

After loading this experimental driver, you should be able to disable BIOS fan control by executing `sudo i8kctl mode1 1`.
Executing `sudo i8kctl mode1 2` should reenable BIOS fan control.

Can you test this and verify that your machine is still working normally (no suspend issues, hotkeys, ...) when BIOS fan control is disabled?

**Comment by tova-cup** [comment]
I did spend some time testing your experimental driver (and thanks a lot of your help ; the results are conclusive !).

First of all, the commands `sudo i8kctl mode1 1` and  `sudo i8kctl mode1 2` throw out the following error :

```
invalid fan mode 1
    while executing
"throw {FANMODE {invalid mode}} "invalid fan mode $mode""
    (class "::hwmon::fanMode" method "setMode" line 10)
    invoked from within
"$fanMode setMode $value"
    (procedure "sensorWrite" line 27)
    invoked from within
"sensorWrite {*}[parseSensorName [lindex $args 0]] [lindex $args 1]"
    (procedure "main" line 15)
    invoked from within
"main {*}$argv"
    (file "/usr/bin/i8kctl" line 242)
```

I did have a quick look at the code, I guess the inputs for the commands should not be integers but `automatic` or `manual` instead.

That said, after running `sudo i8kctl mode1 manual`, **fan control finally works !**
I could set various speed by using `sudo i8kctl fan1 1` with effect this time.
Moreover, I tested multiple scenarios with my PC (suspend, dock usage, shortcut keys, AC/Battery) and **I didn't notice any weird nor broken behaviour.**

I last question though, it seems however that running `i8kmon` doesn't take into account my config file at `/etc/i8kmon.conf`, is this an expected behaviour ?
